lib/streamlit/elements/time_widgets.py

```diff
@@ -33,12 +34,138 @@
 from .form import current_form_id
 from .utils import check_callback_rules, check_session_state_rules
 
+if TYPE_CHECKING:
+    from streamlit.delta_generator import DeltaGenerator
+
+
+TimeValue: TypeAlias = Union[time, datetime, None]
+_DateValue: TypeAlias = Union[date, datetime, None]
+DateValue: TypeAlias = Union[_DateValue, Sequence[_DateValue]]
+DateWidgetReturn: TypeAlias = Union[date, Tuple[()], Tuple[date], Tuple[date, date]]
+
+
+def _parse_date_value(value: DateValue) -> Tuple[List[date], bool]:
+    parsed_dates: List[date]
+    range_value: bool = False
+    if value is None:
+        # Set value default.
+        parsed_dates = [datetime.now().date()]
+    elif isinstance(value, datetime):
+        parsed_dates = [value.date()]
+    elif isinstance(value, date):
+        parsed_dates = [value]
+    elif isinstance(value, (list, tuple)):
+        if not len(value) in (0, 1, 2):
+            raise StreamlitAPIException(
+                "DateInput value should either be an date/datetime or a list/tuple of "
+                "0 - 2 date/datetime values"
+            )
+
+        parsed_dates = [v.date() if isinstance(v, datetime) else v for v in value]
+        range_value = True
+    else:
+        raise StreamlitAPIException(
+            "DateInput value should either be an date/datetime or a list/tuple of "
+            "0 - 2 date/datetime values"
+        )
+    return parsed_dates, range_value
+
+
+def _parse_min_date(
+    min_value: _DateValue,
+    parsed_dates: Sequence[date],
+) -> date:
+    parsed_min_date: date
+    if isinstance(min_value, datetime):
+        parsed_min_date = min_value.date()
+    elif isinstance(min_value, date):
+        parsed_min_date = min_value
+    elif min_value is None:
+        if parsed_dates:
+            parsed_min_date = parsed_dates[0] - relativedelta.relativedelta(years=10)
+        else:
+            parsed_min_date = date.today() - relativedelta.relativedelta(years=10)
+    else:
+        raise StreamlitAPIException(
+            "DateInput min should either be a date/datetime or None"
+        )
+    return parsed_min_date
+
+
+def _parse_max_date(
+    max_value: _DateValue,
+    parsed_dates: Sequence[date],
+) -> date:
+    parsed_max_date: date
+    if isinstance(max_value, datetime):
+        parsed_max_date = max_value.date()
+    elif isinstance(max_value, date):
+        parsed_max_date = max_value
+    elif max_value is None:
+        if parsed_dates:
+            parsed_max_date = parsed_dates[-1] + relativedelta.relativedelta(years=10)
+        else:
+            parsed_max_date = date.today() + relativedelta.relativedelta(years=10)
+    else:
+        raise StreamlitAPIException(
+            "DateInput max should either be a date/datetime or None"
+        )
+    return parsed_max_date
+
+
+@attr.s(auto_attribs=True, slots=True, frozen=True)
+class _DateInputValues:
```

Comment:
I ended up with some extensive refactoring for the date widget, for the purpose of adding types. The most significant change is the introduction and use of this `_DateInputValues` class, the purpose of which I'll explain here.

One bug-or-feature of mypy is that the types declared for a variable are [final and cannot be changed](https://adamj.eu/tech/2021/05/23/python-type-hints-mypy-doesnt-allow-variables-to-change-type/). In the date widget, it turned out the type of `value` was systematically narrowed from the full `DateValue` type to just `Sequence[date]`, requiring assignment to a new variable in order to express the narrowing with types. A similar narrowing was going on for `min_value` and `max_value`. There were also relational consistency checks between these three objects.

The purpose of an instance of `_DateInputValues`, is to represent the final parsed state of these parameters, as they are used in practice. It gathers parsing and validation in such a way that you cannot create an instance of this object without the correct invariances being in place. This means the rest of the code can operate with as narrow types as possible.